src/Common/src/System/Security/Cryptography/Asn1V2.cs

```diff
@@ -0,0 +1,5229 @@
+        public int TagValue => _tagValue;
+
+        private Asn1Tag(byte controlFlags, int tagValue)
+        {
+            _controlFlags = (byte)(controlFlags & ControlMask);
+            _tagValue = tagValue;
+        }
+
+        public Asn1Tag(UniversalTagNumber universalTagNumber, bool isConstructed = false)
+            : this(isConstructed ? ConstructedMask : (byte)0, (int)universalTagNumber)
+        {
+            if (!Enum.IsDefined(typeof(UniversalTagNumber), universalTagNumber))
+            {
+                throw new ArgumentOutOfRangeException(nameof(universalTagNumber), universalTagNumber, null);
+            }
+        }
+
+        public Asn1Tag(TagClass tagClass, int tagValue, bool isConstructed = false)
+            : this((byte)((byte)tagClass | (isConstructed ? ConstructedMask : 0)), tagValue)
+        {
+            if (!Enum.IsDefined(typeof(TagClass), tagClass))
+            {
+                throw new ArgumentOutOfRangeException(nameof(tagClass), tagClass, null);
+            }
+        }
+
+        public Asn1Tag(byte singleByteEncoding)
+        {
+            unsafe
+            {
+                byte* data = &singleByteEncoding;
+
+                if (TryParse(new ReadOnlySpan<byte>(data, 1), out Asn1Tag parsed, out int bytesRead))
+                {
+                    Debug.Assert(bytesRead == 1);
+                    _controlFlags = parsed._controlFlags;
+                    _tagValue = parsed._tagValue;
+                }
+                else
+                {
+                    throw new CryptographicException(SR.Cryptography_Der_Invalid_Encoding);
+                }
+            }
+        }
+
+        public static bool TryParse(ReadOnlySpan<byte> source, out Asn1Tag tag, out int bytesRead)
+        {
+            tag = default(Asn1Tag);
+            bytesRead = 0;
+
+            if (source.IsEmpty)
+                return false;
+
+            bytesRead++;
+            byte first = source[0];
+            uint tagValue = (uint)(first & TagNumberMask);
+
+            if (tagValue == TagNumberMask)
+            {
+                // Multi-byte encoding
+                // T-REC-X.690-201508 sec 8.1.2.4
+                const byte ContinuationFlag = 0x80;
+                const byte ValueMask = ContinuationFlag - 1;
+
+                tagValue = 0;
+                byte current;
+
+                do
+                {
+                    if (source.Length <= bytesRead)
+                    {
+                        bytesRead = 0;
+                        return false;
+                    }
+
+                    current = source[bytesRead];
+                    byte currentValue = (byte)(current & ValueMask);
+                    bytesRead++;
+
+                    // The first byte cannot have the value 0 (T-REC-X.690-201508 sec 8.1.2.4.2.c)
+                    if (currentValue == 0 && tagValue == 0)
+                    {
+                        bytesRead = 0;
```

Comment:
Agree with Morgan: Return `false` instead of throwing if the provided data is invalid.

Reply:
The rules of the `Try` pattern say that you return false for one class of error, but continue to throw on the rest.  Following the rest of our read-from-a-Span work we've done, the pattern best aligns with "if we hit the end of the span, return false, otherwise throw".

https://blogs.msdn.microsoft.com/kcwalina/2005/03/16/design-guidelines-update-exception-throwing/

> When using this pattern, it is important to define the ‘try’ functionality in strict terms. If the member fails for any reason other than the well defined try, then the member must still throw. 

The errors here include:

* Empty input
* Last byte in the span had a continuation marker
* A leading 0-byte was found
* A multi-byte encoding was used when a single byte encoding could have been
* The tag value exceeds (signed) int.MaxValue

If we want to call all of those "the data doesn't represent a valid BER (family) encoded tag" and return false then maybe that's okay.  @KrzysztofCwalina you literally wrote the rules here (on your blog :)), do you have an opinion?

Reply:
```C#
int.TryParse(str, NumberStyles.AllowExponent, CultureInfo.CurrentCulture, out int val);
```

* `null`: return false
* `""`: return false
  * So empty input should return false
* "1g": return false
* "1e": return false
  * So an unresolved continuation marker should return false
* "1e5": return true, out 100,000
* "1e9": return true, out 1,000,000,000
* "1e10": return false
  * So out of range should maybe return false
* "--1": return false
  * So a leading 0 should return false

Supporting values bigger than int.MaxValue would require changing the type of the TagValue property, so that'd be a breaking change, so `return false` makes sense, I guess.

So I'll change it to have no throws.